BrainPortal/app/models/boutiques_cluster_task.rb

```diff
@@ -0,0 +1,354 @@
+    substitutions_by_ids    = JSON.parse(File.read(invoke_json_basename)) # id => val
+    substitutions_by_tokens = descriptor.inputs.map do |input|
+      next nil if input.value_key.blank?
+      value = substitutions_by_ids[input.id]
+      next nil if value.nil?
+      [ input.value_key, value ]
+    end.compact.to_h
+
+    all_ok           = true
+    cbrain_to_ignore = custom['cbrain:ignore_outputs'] || []
+    descriptor.output_files
+      .select { |output| ! cbrain_to_ignore.include?(output.id) }
+      .each do |output|
+      globpath = output.path_template
+      to_strip = output.path_template_stripped_extensions || []
+
+      # Apply substitutions
+      substitutions_by_tokens.each do |key,val|
+        next if val.is_a?(Array) # not supported; what would it mean?
+        val = val.to_s
+        to_strip.each { |str| val.sub!(/\Q#{str}\z/,"") }
+        globpath.gsub!(key, val)
+      end
+
+      paths = Dir.glob(globpath)
+      if output.optional && paths.empty?
+        self.addlog("Skipped optional missing output file '#{globpath}' for output #{output.id}")
+        next
+      end
+
+      paths.each do |path|
+        if ! path_is_in_workdir?(path) # this also checks the existence
+          self.addlog("Output file is missing or outside of task work directory: #{path}")
+          all_ok = false
+        end
+
+        # Get name and filetype
+        self.addlog("Attempting to save result file #{path}")
+        name = File.basename(path)
+        userfile_class   = Userfile.suggested_file_type(name)
+        userfile_class ||= ( File.directory?(path) ? FileCollection : SingleFile )
+
+        # Add a run ID to the file name, to make sure the file doesn't exist.
+        name.sub!( /(\.\w+(\.gz|\.z|\.bz2|\.zip)?)?\z/i )  { |ext| "-#{self.run_id}" + ext }
+
+        # Save the file (possible overwrite if race condition)
+        outfile = safe_userfile_find_or_new(userfile_class, :name => name)
+
+        unless outfile.save
+          messages = outfile.errors.full_messages.join("; ")
+          self.addlog("Failed to save file #{path} as #{name}")
+          self.addlog(messages) if messages.present?
+          all_ok = false
+          next
+        end
+
+        # Transfer content to DataProvider
+        outfile.cache_copy_from_local_file(path)
+        params["_cbrain_output_#{output.id}"] ||= []
+        params["_cbrain_output_#{output.id}"]  << outfile.id
+        self.addlog("Saved result file #{name}")
+
+        # Add provenance logs
+        all_file_input_ids = descriptor.file_inputs.map do |input|
+          invoke_params[input.id]
+        end.compact.uniq
+        parents = Userfile.where(:id => all_file_input_ids).to_a
+        self.addlog_to_userfiles_these_created_these(parents, [outfile]) if parents.present?
+
+        # If there is only one input file, we move the output under it
+        outfile.move_to_child_of(parents[0]) if parents.size == 1
```

Comment:
If we have multiple files but only one is mandatory, can we move the output under it too ?

Reply:
It's a good idea. I'll do it.